sharktank/sharktank/models/clip/clip.py

```diff
@@ -0,0 +1,502 @@
+# Copyright 2021 The OpenAI Team Authors and The HuggingFace Team.
+# Copyright 2024 Advanced Micro Devices, Inc.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from typing import Any, Optional, Tuple, Union
+import torch
+from torch import nn
+import transformers
+
+# TODO: port _prepare_4d_attention_mask and _create_4d_causal_attention_mask to sharktank
+from transformers.modeling_attn_mask_utils import (
+    _prepare_4d_attention_mask,
+    _create_4d_causal_attention_mask,
+)
+from collections import OrderedDict
+
+from ...layers import BaseLayer, LinearLayer, LayerNorm, TokenEmbeddingLayer
+from ... import ops
+from ...types.theta import Theta, Dataset
+from ...types.tensors import DefaultPrimitiveTensor
+from ...layers.configs import ClipTextConfig
+from ...layers.activations import ACT2FN
+
+
+class ClipTextEmbeddings(nn.Module):
+    def __init__(self, theta: Theta, config: ClipTextConfig):
+        super().__init__()
+        self.token_embedding = TokenEmbeddingLayer(
+            theta=theta("token_embedding"), dtype=None
+        )
+        self.position_embedding = TokenEmbeddingLayer(
+            theta=theta("position_embedding"), dtype=None
+        )
+
+        # position_ids (1, len position emb) is contiguous in memory and exported when serialized
+        self.register_buffer(
+            "position_ids",
+            torch.arange(config.max_position_embeddings).expand((1, -1)),
```

Comment:
You should use `unsqueeze` instead of `expand`

Reply:
I forgot to mention in the PR description that this is the initial port of the model and I did not attempt to optimize anything. The main goal was to put it under test. I would rather do these modifications later as they would be tracked in different commits and will be clear what changed compared to the original.

Reply:
If you're directly referencing the huggingface code for implementation, do you want to link it in a comment?

Reply:
I added a comment at the top of the file.